discord.go

```diff
@@ -50,80 +50,104 @@ func discordInit() {
-		webhooks, err := sess.ChannelWebhooks(Integrations.Discord.ChannelID)
+		webhooks, err = sess.ChannelWebhooks(Integrations.Discord.ChannelID)
 		if err != nil {
 			Log.Println("Error getting Discord webhooks:", err)
-			return
-		}
-		for _, wh := range webhooks {
-			if wh.Name == "Devzat" {
-				webhook = wh
-			}
-		}
-		if webhook == nil {
-			webhook, err = sess.WebhookCreate(Integrations.Discord.ChannelID, "Devzat", "")
-			if err != nil {
-				Log.Println("Error creating a Discord webhook:", err)
-				return
-			}
+			webhooks = make([]*discordgo.Webhook, 0, maxWebhookCount)
 		}
 	}
 	DiscordChan = make(chan DiscordMsg, 100)
-	editsInLastMinute := 0 // discord allows for 30 webhook edits per minute: https://twitter.com/lolpython/status/967621046277820416
 	go func() {
-		overloading := false
+	nextMsg:
 		for msg := range DiscordChan {
-			sendingTimeStart := time.Now()
 			txt := strings.ReplaceAll(msg.msg, "@everyone", "@\\everyone")
-			if Integrations.Discord.CompactMode || overloading {
-				var toSend string
-				if msg.senderName == "" {
-					toSend = strings.ReplaceAll(stripansi.Strip("["+msg.channel+"] "+txt), `\n`, "\n")
-				} else {
-					toSend = strings.ReplaceAll(stripansi.Strip("["+msg.channel+"] **"+msg.senderName+"**: "+txt), `\n`, "\n")
+			if Integrations.Discord.CompactMode {
+				sendDiscordCompactMessage(msg, txt, sess)
+			} else {
+				avatar := createDiscordImage(msg.senderName)
+				avatarHash := shasum(avatar)
+				var webhook *discordgo.Webhook
+
+				// find the webhook for this user
+				for _, wh := range webhooks {
+					if wh.Name == avatarHash {
+						webhook = wh
+						break
+					}
 				}
-				_, err = sess.ChannelMessageSend(Integrations.Discord.ChannelID, toSend)
-				if err != nil {
-					Log.Println("Error sending Discord message:", err)
+				// delete unused webhooks if there are too many and a new one is needed
+				// (discord can only have 15 webhooks per channel)
+				if webhook == nil && len(webhooks) > maxWebhookCount {
+					// generate a list of all users in all rooms
+					users := make([]string, 0, maxWebhookCount)
+					for _, room := range Rooms {
+						for _, user := range room.users {
+							users = append(users, user.Name)
+						}
+					}
+					users = append(users, "", Devbot)
+					// if there are more users than webhooks we are recreating webhooks all the time which would get us
+					// rate limited by Discord, so just switch to compact mode
+					// TODO: AFK detection?
+					if len(users) >= maxWebhookCount {
+						sendDiscordCompactMessage(msg, txt, sess)
+						continue
+					}
+					// find a webhook that is not in use
+					for _, wh := range webhooks {
+						found := false
+						for _, user := range users {
+							if wh.Name == shasum(createDiscordImage(user)) { // generating all the images is cashed
+								found = true
+								break
+							}
+						}
+						if !found {
+							err = sess.WebhookDelete(wh.ID, discordgo.WithRetryOnRatelimit(false))
```

Comment:
wh will always be the last item in the webhooks list, right? is this intentional? also if this is the intended behavior, you no longer need a loop to find the index of this webhook since you know it's the last one.